### ISSUE TITLE
NotoSansTamilUI: Transformations Filter applies twice to composite glyphs

### ISSUE THREAD

**Issue by anthrotype** [googlei18n/noto-source#84]
While working on https://github.com/googlei18n/fontmake/issues/275, I noticed a problem with the way we use Glyphs.app's "Filter" custom parameter in NotoSansTamilUI.

The filter in question is a "Transformations" that is meant to shift all glyphs vertically by 51 units:

```
Transformations;OffsetY:51;Origin:4;
```

`Filter` custom parameters by default apply to all glyphs (unless one specifies "include" or "exclude" values), and are triggered, as the Handbook says, "after decomposition of compound glyphs".

Now, this works as intended when exporting OTF (CFF) instances, where all compound glyphs are always decomposed on export.

However there is a problem when we export TTF instances, because here compound glyphs are not decomposed by default on export, as `glyf` table can handle these; this means that the "Transformations" filter, being applied to all glyphs (including compounds) will end up being applied multiple times for the composite glyphs, because the transformation will occur both in each of the base glyphs used as components, and again for the composite glyph as a whole...

Compare the highlighted subscript two glyph (a composite glyph of the superscript two) in the OTF: 

<img width="238" alt="screenshot 2017-08-10 11 15 07" src="https://user-images.githubusercontent.com/6939968/29166167-39f51d1c-7dbd-11e7-9bdc-488a56687d6a.png">

And in the TTF:
<img width="238" alt="screenshot 2017-08-10 11 15 16" src="https://user-images.githubusercontent.com/6939968/29166169-3ce1401e-7dbd-11e7-8b1b-4ae2ec0e8205.png">

The latter appears to be shifted twice by a doubled amount.

So this means the same generic filter value can't work for producing both OTF and TTF.

To produce the same effect for both output formats, we can do two things:

1) decompose all glyphs also for the generated TTF (but this will increase its file size); Glyphs has a "Decompose glyphs" custom parameter which takes a list of compound glyphs to be decomposed at export time;

2) Or turn the `Filter` into a `PreFilter`, which is another kind of custom parameter that, unlike `Filter`, is applied *before* decomposition of compound glyphs; _but_ we need to apply this pre-filter only to non-composite glyphs (i.e. containing actual paths, not references to other glyphs). This way, the composite glyphs will also be indirectly transformed, because their respective base glyphs were; but they themselves won't be transformed as such, because we exclude them from the filter. To do that, we can append to the "Transformations..." string above, either an ‘include:’ list of  space- or comma-separated names of non-compound glyphs, or which is the same an ‘exclude:’ list with the names of the all the composite glyphs.

The (2) approach is preferable, however we need to check if there are glyphs containing mixed contours and components, which should be decomposed before applying the filter, or else their components will have a wrong transform. Adding only these to the "Decompose Glyphs" custom parameter should work.

**Comment by behdad** [comment]
What's the resolution here?

**Comment by anthrotype** [comment]
> What's the resolution here?

The resolution here is that we should use a "PreFilter" instead of a "Filter" filter, and only include non-composite glyphs in it.
But there's a few things we need before we can do this.

1) glyphLib needs to know about PreFilter custom parameters; currently it only reads Filter (https://github.com/googlei18n/glyphsLib/issues/214). It should be relatively simple to extend -- only that glyphsLib master branch is blocked by the `classtree` rewrite...

2) ufo2ft needs to handle both Pre- and regular Filters, however the problem here is that the decomposition of composite glyphs (as well as overlap removal and quadratic conversion) currently happens inside fontmake *before* and independently of ufo2ft; so by the time the fonts reach ufo2ft, it's too late to apply a PreFilter (i.e one that occurs before decomposition, and thus logically also before overlap removal and quadratic conversion).

I was already thinking for a while of moving the composite glyphs' decomposition and the overlap removal steps from fontmake to ufo2ft, with the latter as the compiler API and the former as its command line front end. As for the simple (non-interpolatable) cubic-to-quadratic conversion, this is already supported by ufo2ft. Having the decomposition and overlap removal steps performed by ufo2ft would allow it to apply the filters in the right order (pre- or not). In this scenario, the interpolatable/variable output would still be done by fontmake rather than ufo2ft's, as the latter only does one single ufo at a time (but that could also be changed).

Another option, which is tempting, is that we let glyphsLib apply the filters' transformations on the generated UFO instances themselves, instead of leaving it to the ufo2ft compiler to apply them. We would call glyph.move((x, y)) and we'd be basically done. I don't know... It would mean we would only support filters with *.glyphs input sources, and not with UFO sources; but it's true that this idea of an export-time filter come from Glyphs.app originally.

**Comment by moyogo** [comment]
We had a use case for a prefilter in UFOs.
Designers wanted to have the outlines in a certain position to make consistent design easier (in our case for a connected script font), only the have them offset at build time.

**Comment by anthrotype** [comment]
I just tried to re-export TTFs for NotoSansTamil UI instances using the latest Glyphs.app beta 1093.
The UI instances contain the following custom parameter

```
(
        {
        Filter = "Transformations;OffsetY:51;Origin:4;";
    }
)
```

With older Glyphs.app versions, the composite glyphs would be shifted up twice -- as I explained in the original post above.

With the current beta, it seems that they are _not_ being shifted twice...

I implemented the Transformations filter in ufo2ft following the old Glyphs.app behavior... 😞 
I was planning to add PreFilter with only the contour-glyphs to fix the above issue, but seems like Glyphs.app behavior has now changed.

@schriftgestalt can you please explain how composite glyphs are supposed to be affected by the Transformations export filter?
What about a filter has include statement which contains some composite glyphs? Do you ignore an explicit request from the user?
And how are mixed contour/component glyphs treated in this regard?

**Comment by anthrotype** [comment]
I made some tests, and this is what I can see:

- when a glyph contains components and it is included in the transformation, if the base glyph for that component is also included in the list of glyphs to be transformed, then the base glyph is transformed only once, and its component reference in the composite glyphs is not transformed

- if the component's base glyph is not also included, then the component itself is transformed, and the base glyph is left untouched.

@schriftgestalt can you confirm this is how it works?

**Comment by schriftgestalt** [comment]
Your observations are spot on.

**Comment by anthrotype** [comment]
not sure what you mean by that "compensation of the movement of the base glyphs".
I am simply not transforming a component if its base glyph (which can be another composite glyph) is also included in the transformation.

**Comment by schriftgestalt** [comment]
Try a component that is 180 degrees rotated.

**Comment by anthrotype** [comment]
I will try, but still I don’t see why the kind of transformation should make any difference.

**Comment by anthrotype** [comment]
Oh... I see! E.g. if user says “shift up” but the component is flipped upside down, you have to shift it down, and things like that.

**Comment by anthrotype** [comment]
@schriftgestalt I tried to apply a "Filter = Transformations;OffsetY:200" to a composite glyph flipped upside down (ScaleY -100), as well as to one rotated 180 degrees.

When I export the instance TTF from Glyphs.app, I see that the two composite glyphs have been decomposed, and therefore correctly shifted up by 200 units.

If instead of Filter, I use a PreFilter, the result is that two composite glyphs are still decomposed, but they are now shifted _down_ by 200 units.

Besides the decomposition of transformed glyphs -- there must be some custom parameter to turn this off (I forgot which one) --, I don't see any special "compensation of the movement of the base glyph" that you mentioned above...

**Comment by anthrotype** [comment]
this is the test glyphs file I used:
[test_transform.zip](https://github.com/googlei18n/noto-source/files/1530589/test_transform.zip)

**Comment by anthrotype** [comment]
ok, I had to turn on "Keep Transformed Components". Now I see what you mean.

that's another issue: Glyphs.app defaults to decomposing transformed components, while fontmake keeps them by default for TTFs https://github.com/googlei18n/fontmake/issues/253

**Comment by anthrotype** [comment]
I checked NotoSansTamil which is the only font that uses the Transformations filter, and it seems like no component has any scale or slant or flip, just plain offsets.

So I'm leaning towards releasing this first version of the ufo2ft filter, which doesn't take into account components with complex transformations, and add support for that later when needed.

**Comment by anthrotype** [comment]
also, I believe that to be complete, this Tranformations filter should also nullify (invert) the transformation on components that had their base glyph transformed but whose composite glyphs they belong was not included in the set to be transformed.
Glyphs.app doesn't seem to do that.

wow, this is getting complicated...

**Comment by anthrotype** [comment]
ok I think I've found an algorithm: when about to transform a component, check if the base glyph was already transformed; if so, then check if the component's current 6-float tuple transformation starts with (1, 0, 0, 1), then don't transform the component again as the base already took care of it; otherwise, if the component's current transformation is not a simple offset, then multiply the component's transformation matrix with the _inverse_ of the filter's transformation matrix (i.e. the "compensation" step Georg was talking about), and then proceed to apply the filter's matrix on top that.
